<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Non-local prior based iterative variable selection for GWAS data, or other high-dimensional data — GWASinlps • GWASinlps</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Non-local prior based iterative variable selection for GWAS data, or other high-dimensional data — GWASinlps"><meta property="og:description" content="GWASinlps performs variable selection with data from Genome-wide association studies (GWAS), or other high-dimensional data, combining in an iterative framework, the computational efficiency of the screen-and-select variable selection approach based on some association learning and the parsimonious uncertainty quantification provided by the use of non-local priors as described in Sanyal et al. (2019)."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GWASinlps</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/nilotpalsanyal/GWASinlps/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Non-local prior based iterative variable selection for GWAS data, or other high-dimensional data</h1>
    
    <div class="hidden name"><code>GWASinlps.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>GWASinlps</code> performs variable selection with data from Genome-wide association studies (GWAS), or other high-dimensional data, combining in an iterative framework, the computational efficiency of the screen-and-select variable selection approach based on some association learning and the parsimonious uncertainty quantification provided by the use of non-local priors as described in Sanyal et al. (2019).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">GWASinlps</span><span class="op">(</span></span>
<span>  <span class="va">y</span>, </span>
<span>  <span class="va">x</span>, </span>
<span>  family <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"normal"</span>,<span class="st">"binomial"</span><span class="op">)</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rigorous"</span>,<span class="st">"quick"</span><span class="op">)</span>,</span>
<span>  cor_xy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  mmle_xy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  prior <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mom"</span>, <span class="st">"imom"</span>, <span class="st">"zellner"</span>, <span class="st">"horseshoe"</span><span class="op">)</span>, </span>
<span>  <span class="va">tau</span>, </span>
<span>  priorDelta <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/mombf/man/msPriorSpec-class.html" class="external-link">modelbbprior</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>, </span>
<span>  <span class="va">k0</span>, </span>
<span>  <span class="va">m</span>, </span>
<span>  <span class="va">rxx</span>, </span>
<span>  nskip <span class="op">=</span> <span class="fl">3</span>, </span>
<span>  niter <span class="op">=</span> <span class="fl">2000</span>, </span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>  seed <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>  tau.hs.method <span class="op">=</span> <span class="st">"halfCauchy"</span>, </span>
<span>  sigma.hs.method <span class="op">=</span> <span class="st">"Jeffreys"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>y</dt>
<dd><p>The vector of continuous or binary response (phenotype). Binary response values must be 0 or 1.</p></dd>

  <dt>x</dt>
<dd><p>The design matrix with subjects in rows and independent variables (SNPs) in columns. Missing values are not accepted currently.</p></dd>

  <dt>family</dt>
<dd><p><code>"normal"</code> for continuous data, <code>"binomial"</code> for binary data (logit link is used).</p></dd>

  <dt>method</dt>
<dd><p>Applies only when <code>family="binomial"</code>. The <code>rigorous</code> method uses logistic regression based analysis which is theoretically appropriate but can be slow. The <code>quick</code> method uses a curious combination of linear model and logistic regression based analyses and is considerably faster. See Details.</p></dd>

  <dt>cor_xy</dt>
<dd><p>Used only when <code>family="normal"</code>. Vector of (Pearson) correlation coefficients of <code>y</code> with the individual columns of <code>x</code>.</p></dd>

  <dt>mmle_xy</dt>
<dd><p>Used only when <code>family="binomial"</code>. Vector of maximum marginal likelihood estimates of the regression parameters corresponding to the <code>x</code> variables (SNPs). These are obtained from GLM fits of <code>y</code> with the individual columns of <code>x</code> including an intercept.</p></dd>

  <dt>prior</dt>
<dd><p><code>"mom"</code> for pMOM prior, <code>"imom"</code> for piMOM prior, <code>"zellner"</code> for Zellner's g-prior, <code>"horseshoe"</code> for horseshoe prior. For <code>family="binomial"</code>, <code>"zellner"</code> considers group Zellner prior and <code>"horseshoe"</code> is not available.</p></dd>

  <dt>tau</dt>
<dd><p>The value of the scale parameter tau of the non-local prior.</p></dd>

  <dt>priorDelta</dt>
<dd><p>Prior for model space. Defaults to <code>modelbbprior(1,1)</code>, which is beta-binomial(1,1) prior.</p></dd>

  <dt>k0</dt>
<dd><p>GWASinlps tuning parameter denoting the number of leading SNPs (see Details).</p></dd>

  <dt>m</dt>
<dd><p>GWASinlps tuning parameter, denoting the maximum number of SNPs to be selected.</p></dd>

  <dt>rxx</dt>
<dd><p>GWASinlps tuning parameter denoting the correlation threshold to determine leading sets (see References).</p></dd>

  <dt>nskip</dt>
<dd><p>GWASinlps tuning parameter denoting the maximum allowed count of skipping an iteration that does not select any variable (SNP) (see References).</p></dd>

  <dt>niter</dt>
<dd><p>Number of MCMC iterations for non-local prior based Bayesian variable selection. Defaults to 2000.</p></dd>

  <dt>verbose</dt>
<dd><p>If TRUE, prints result from the iterations progressively. FALSE by default.</p></dd>

  <dt>seed</dt>
<dd><p>For reproducibility. If provided, the random seed is set to this value at the beginning of the function.</p></dd>

  <dt>tau.hs.method</dt>
<dd><p>Necessary only when <code>prior="horseshoe"</code>. See <code>horseshoe</code> function reference.</p></dd>

  <dt>sigma.hs.method</dt>
<dd><p>Necessary only when <code>prior="horseshoe"</code>. See <code>horseshoe</code> function reference.</p></dd>

</dl></div>
    <div id="details">
    <h2>Details</h2>
    <p>The GWASinlps method selects variables (SNPs) iteratively.</p>
<div class="section">
<h3 id="for-continuous-response">For continuous response<a class="anchor" aria-label="anchor" href="#for-continuous-response"></a></h3>


<p>For continuous response (phenotype), the procedure starts with an initial set of independent variables (SNPs), a design matrix (SNP genotype matrix) <code>x</code> and a response (phenotype) vector <code>y</code>.</p>
<p>- An iteration proceeds by determining the <code>k0</code> <em>leading SNPs</em> having the highest association with <code>y</code>. The measure of association is the absolute value of the Pearson's correlation coefficient <code>cor_xy</code>. These <code>k0</code> leading SNPs, in turn, determine <code>k0</code> <em>leading sets</em>, where each leading set consists of all SNPs with absolute correlation coefficient more than or equal to <code>rxx</code> with the correspondng leading SNP.</p>
<p>- Within each leading set, non-local prior based Bayesian variable selection for linear models is performed (using package <span class="pkg">mombf</span>). The variables (SNPs) appearing in the highest posterior probability model (HPPM) are considered selected in the current iteration. Note that a single variable (SNP) can be selected from multiple leading sets. The selected variables (SNPs) are regressed out from <code>y</code> using <code>lm</code> fit. The variables (SNPs) that are included in one or more of the <em>leading sets</em> but do not appear in any HPPM are dropped from further analysis.</p>
<p>- With updated <code>y</code> and variable (SNP) set, next iteration proceeds similarly. And so on like this. The procedure continues until the stopping point, which is determined by the GWASinlps tuning parameters <code>m</code>, <code>rxx</code>, and <code>nskip</code>, is reached. For more details, see References.</p>
</div>


<div class="section">
<h3 id="for-binary-response">For binary response<a class="anchor" aria-label="anchor" href="#for-binary-response"></a></h3>


<p>For binary response (phenotype), the procedure starts with an initial set of variables (SNPs), a design matrix (SNP genotype matrix) <code>x</code> and a binary response (phenotype) vector <code>y</code>. If <code>method="rigorous"</code>,</p>
<p>- The first iteration proceeds by determining the <code>k0</code> <em>leading SNPs</em> having the highest association with <code>y</code>. The measure of association is the absolute value of the maximum marginal likelihood estimate <code>mmle_xy</code>. These <code>k0</code> leading SNPs, in turn, determine <code>k0</code> <em>leading sets</em>, where each leading set consists of all SNPs with absolute correlation coefficient more than or equal to <code>rxx</code> with the correspondng leading SNP.</p>
<p>- Within each leading set, non-local prior based Bayesian variable selection for logistic regression model is performed (using package <span class="pkg">mombf</span>). The variables (SNPs) appearing in the HPPM are considered selected in the first iteration. Note that a single variable (SNP) can be selected from multiple leading sets. The variables (SNPs) which are included in one or more <em>leading sets</em> but do not appear in any HPPM are dropped from further analysis. After this, the selected variables (SNPs) are accounted for by including them in <code>glm</code> fits of <code>y</code> with each of the remaining variables (SNPs). The <code>glm</code> coefficients of the remaining variables, thus obtained, reflect their contribution in presence of the selected variables (SNPs) of the first iteration.</p>
<p>- Considering the absolute values of these <code>glm</code> coefficients as the measure of association, we proceed with the second iteration with updated variable (SNP) set. And so on in this manner. The procedure continues until the stopping point, which is determined by the GWASinlps tuning parameters <code>m</code>, <code>rxx</code>, and <code>nskip</code>, is reached.</p>
<p>If <code>method="quick"</code>, the procedure is similar to above except at the following points. In this method, non-local prior based Bayesian variable selection using logistic regression model is performed until one or more variables (SNP) are selected in an iteration. Until a variable is selected, there is no need to account for anything, so the initial maximum marginal likelihood estimates continue to be used. After the first selections (if any) are made, a <code>glm</code> fit of <code>y</code> on the selected variables is performed and the deviance residuals are computed. In the subsequent iterations, considering these (continuous) deviance residuals as response, non-local prior based Bayesian variable selection for linear models is performed till the stopping point is reached.</p>
</div>


<div class="section">
<h3 id=""></h3>
<p>For horseshoe prior, package <span class="pkg">horseshoe</span> is used.</p>
</div>


    </div>
    <div id="value">
    <h2>Value</h2>
    

<p>A list containing</p>
<dl><dt>selected </dt>
<dd><p>Vector with names of the GWASinlps selected variables (SNPs) in the order they were selected.</p></dd>

    <dt>selected_iterwise </dt>
<dd><p>List with selected variables (SNPs) from each iteration.</p></dd>

</dl></div>
    <div id="references">
    <h2>References</h2>
    <p>Sanyal et al. (2019), "GWASinlps: Non-local prior based iterative SNP selection tool for genome-wide association studies". Bioinformatics, 35(1), 1-11.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="nlps.html">nlpsLM</a></code>, <code><a href="nlps.html">nlpsGLM</a></code>, <code><a href="https://rdrr.io/pkg/mombf/man/modelSelection.html" class="external-link">modelSelection</a></code>, <code><a href="https://rdrr.io/pkg/horseshoe/man/horseshoe.html" class="external-link">horseshoe</a></code></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Nilotpal Sanyal &lt;nilotpal.sanyal@gmail.com&gt;</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">n</span> <span class="op">=</span> <span class="fl">200</span></span></span>
<span class="r-in"><span><span class="va">p</span> <span class="op">=</span> <span class="fl">1000</span></span></span>
<span class="r-in"><span><span class="va">m</span> <span class="op">=</span> <span class="fl">10</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Generate design matrix (genotype matrix)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span> </span></span>
<span class="r-in"><span><span class="va">f</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span> <span class="va">p</span>, <span class="fl">.1</span>, <span class="fl">.2</span> <span class="op">)</span> <span class="co"># simulate minor allele frequency</span></span></span>
<span class="r-in"><span><span class="va">x</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span> nrow <span class="op">=</span> <span class="va">n</span>, ncol <span class="op">=</span> <span class="va">p</span> <span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="va">p</span></span></span>
<span class="r-in"><span><span class="kw">for</span><span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">p</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">x</span><span class="op">[</span>,<span class="va">j</span><span class="op">]</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Binomial.html" class="external-link">rbinom</a></span><span class="op">(</span> <span class="va">n</span>, <span class="fl">2</span>, <span class="va">f</span><span class="op">[</span><span class="va">j</span><span class="op">]</span> <span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Generate true effect sizes</span></span></span>
<span class="r-in"><span><span class="va">causal_snps</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span> <span class="fl">1</span><span class="op">:</span><span class="va">p</span>, <span class="va">m</span> <span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">beta</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span> <span class="fl">0</span>, <span class="va">p</span> <span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">beta</span><span class="op">[</span><span class="va">causal_snps</span><span class="op">]</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="va">m</span>, mean <span class="op">=</span> <span class="fl">0</span>, sd <span class="op">=</span> <span class="fl">2</span> <span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Generate continuous (phenotype) data</span></span></span>
<span class="r-in"><span><span class="va">y</span> <span class="op">=</span> <span class="va">x</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span> <span class="va">beta</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="va">n</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span> </span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Fix scale parameter tau </span></span></span>
<span class="r-in"><span><span class="va">tau</span> <span class="op">=</span> <span class="fl">0.2</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># GWASinlps analysis</span></span></span>
<span class="r-in"><span><span class="va">inlps</span> <span class="op">=</span> <span class="fu">GWASinlps</span><span class="op">(</span><span class="va">y</span>, <span class="va">x</span>, family<span class="op">=</span><span class="st">"normal"</span>, prior<span class="op">=</span><span class="st">"mom"</span>, tau<span class="op">=</span><span class="va">tau</span>, k0<span class="op">=</span><span class="fl">1</span>, m<span class="op">=</span><span class="fl">50</span>, </span></span>
<span class="r-in"><span>        rxx<span class="op">=</span><span class="fl">0.2</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> =================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of selected variables: 17</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time taken: 0 min</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> =================================</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span> <span class="st">"GWASinlps selected"</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">inlps</span><span class="op">$</span><span class="va">selected</span><span class="op">)</span>, <span class="st">"SNPs with"</span>, </span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">inlps</span><span class="op">$</span><span class="va">selected</span>, <span class="va">causal_snps</span><span class="op">)</span><span class="op">)</span>, <span class="st">"true positive(s) and"</span>,</span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="va">causal_snps</span>, <span class="va">inlps</span><span class="op">$</span><span class="va">selected</span><span class="op">)</span><span class="op">)</span>, <span class="st">"false negative(s) out  </span></span></span>
<span class="r-in"><span><span class="st">    of a pool of"</span>, <span class="va">p</span>, <span class="st">"SNPs with data from"</span>, <span class="va">n</span>, <span class="st">"persons."</span> <span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GWASinlps selected 17 SNPs with 10 true positive(s) and 0 false negative(s) out  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     of a pool of 1000 SNPs with data from 200 persons.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Compare with LASSO</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://glmnet.stanford.edu" class="external-link">glmnet</a></span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: Matrix</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loaded glmnet 4.1-4</span>
<span class="r-in"><span><span class="va">fit.cvlasso</span> <span class="op">=</span> <span class="fu"><a href="https://glmnet.stanford.edu/reference/cv.glmnet.html" class="external-link">cv.glmnet</a></span><span class="op">(</span> <span class="va">x</span>, <span class="va">y</span>, alpha <span class="op">=</span> <span class="fl">1</span> <span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">l.min</span> <span class="op">=</span> <span class="va">fit.cvlasso</span> <span class="op">$</span><span class="va">lambda.min</span> <span class="co"># lambda that gives minimum cvm</span></span></span>
<span class="r-in"><span><span class="va">l.1se</span> <span class="op">=</span> <span class="va">fit.cvlasso</span> <span class="op">$</span><span class="va">lambda.1se</span>  <span class="co"># largest lambda such that error is </span></span></span>
<span class="r-in"><span>                                 <span class="co"># within 1 se of the minimum</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">lasso_min</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/stats/coef.html" class="external-link">coef</a></span><span class="op">(</span> <span class="va">fit.cvlasso</span>, s <span class="op">=</span> <span class="va">l.min</span> <span class="op">)</span> <span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span> <span class="op">!=</span> <span class="fl">0</span> <span class="op">)</span>  </span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span> <span class="st">"LASSO with lambda.min selected"</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">lasso_min</span><span class="op">)</span>, <span class="st">"SNPs with"</span>, </span></span>
<span class="r-in"><span>     <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">lasso_min</span>, <span class="va">causal_snps</span><span class="op">)</span><span class="op">)</span>, <span class="st">"true positives and"</span>,</span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="va">causal_snps</span>, <span class="va">inlps</span><span class="op">$</span><span class="va">selected</span><span class="op">)</span><span class="op">)</span>, <span class="st">"false negative(s)."</span> <span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> LASSO with lambda.min selected 63 SNPs with 10 true positives and 0 false negative(s).</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">lasso_1se</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/stats/coef.html" class="external-link">coef</a></span><span class="op">(</span> <span class="va">fit.cvlasso</span>, s <span class="op">=</span> <span class="va">l.1se</span> <span class="op">)</span> <span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span> <span class="op">!=</span> <span class="fl">0</span> <span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span> <span class="st">"LASSO with lambda.1se selected"</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">lasso_1se</span><span class="op">)</span>, <span class="st">"SNPs with"</span>, </span></span>
<span class="r-in"><span>     <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">lasso_1se</span>, <span class="va">causal_snps</span><span class="op">)</span><span class="op">)</span>, <span class="st">"true positives and"</span>,</span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="va">causal_snps</span>, <span class="va">inlps</span><span class="op">$</span><span class="va">selected</span><span class="op">)</span><span class="op">)</span>, <span class="st">"false negative(s)."</span> <span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> LASSO with lambda.1se selected 24 SNPs with 9 true positives and 0 false negative(s).</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Nilotpal Sanyal.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

